src/control/lib/spdk/src/nvme_control_common.c

```diff
@@ -428,41 +470,46 @@ _collect(struct ret_t *ret, data_copier copy_data, pci_getter get_pci,
 			goto fail;
 		}
 
-		ctrlr_tmp->nss = NULL;
-		ctrlr_tmp->stats = NULL;
-		ctrlr_tmp->next = NULL;
-
 		cdata = spdk_nvme_ctrlr_get_data(ctrlr_entry->ctrlr);
 
 		rc = copy_data(ctrlr_tmp, cdata);
 		if (rc != 0)
 			goto fail;
 
-		rc = spdk_pci_addr_fmt(ctrlr_tmp->pci_addr,
-				       sizeof(ctrlr_tmp->pci_addr),
+		ctrlr_tmp->pci_addr = calloc(1, SPDK_NVMF_TRADDR_MAX_LEN + 1);
+		if (!ctrlr_tmp->pci_addr) {
+			rc = -ENOMEM;
+			goto fail;
+		}
+
+		rc = spdk_pci_addr_fmt(ctrlr_tmp->pci_addr, SPDK_NVMF_TRADDR_MAX_LEN,
 				       &ctrlr_entry->pci_addr);
 		if (rc != 0) {
 			rc = -NVMEC_ERR_PCI_ADDR_FMT;
 			goto fail;
 		}
+		if (strlen(ctrlr_tmp->pci_addr) == 0) {
+			rc = -NVMEC_ERR_PCI_ADDR_FMT;
+			goto fail;
+		}
+
+		/* Populate numa socket id & pci device type */
 
 		pci_dev = get_pci(ctrlr_entry->ctrlr);
 		if (!pci_dev) {
 			rc = -NVMEC_ERR_GET_PCI_DEV;
 			goto fail;
 		}
 
-		/* populate numa socket id & pci device type */
 		ctrlr_tmp->socket_id = get_socket_id(pci_dev);
-		written = snprintf(ctrlr_tmp->pci_type,
-				   sizeof(ctrlr_tmp->pci_type), "%s",
-				   spdk_pci_device_get_type(pci_dev));
-		if (written >= sizeof(ctrlr_tmp->pci_type)) {
-			rc = -NVMEC_ERR_CHK_SIZE;
-			free(pci_dev);
+
+		pci_type = spdk_pci_device_get_type(pci_dev);
+		free(pci_dev);
+		ctrlr_tmp->pci_type = strndup(pci_type, strlen(pci_type));
```

Comment:
Unsafe to invoke strlen to figure out the length for strndup. Is there a sensible maximum for this string?

Reply:
done